src/main/network/relay/mod.rs

```diff
@@ -1 +1,219 @@
+use std::net::Ipv4Addr;
+use std::sync::{Arc, Weak};
+
+use atomic_refcell::AtomicRefCell;
+
+use crate::core::work::task::TaskRef;
+use crate::core::worker::Worker;
+use crate::cshadow as c;
+use crate::host::host::Host;
+use crate::network::packet::PacketStatus;
+use crate::network::relay::token_bucket::TokenBucket;
+use crate::network::{Packet, PacketDevice};
+use crate::utility::{Magic, ObjectCounter};
+use shadow_shim_helper_rs::simulation_time::SimulationTime;
+
 mod token_bucket;
+
+/// Specifies a throughput limit when forwarding packets.
+pub enum RateLimit {
+    BytesPerSecond(u64),
+    Unlimited,
+}
+
+/// Relays packets between two packet devices, optionally enforcing a bandwidth
+/// rate limit.
+pub struct Relay {
+    magic: Magic<Self>,
+    _counter: ObjectCounter,
+    // Internals in an Arc so that we can schedule tasks that refer back to it.
+    internal: Arc<AtomicRefCell<RelayInternal>>,
```

Comment:
Yeah, I also have a mild preference for this. This is something that's evolved/evolving a bit in the codebase; I think some existing code encapsulates `Arc` etc so make the C APIs simpler, or otherwise hide the implementation details of which wrappers we were using etc. Even then though it'd probably be better to use a type alias or newtype like `RelayRc` to make it clearer that it's a ref-counted object.